File: mod/game/debug.lua
local debug = {}

local Str = require("mod/tool/string")

debug.Init = function(mod)
    if debug.active == nil then debug.active = true end
    if debug.counter == nil then debug.counter = 0 end
    if debug.Messages == nil then debug.Messages = {} end
    if debug.Variables == nil then debug.Variables = {} end
    if debug.EntitiesWithMessages == nil then debug.EntitiesWithMessages = {} end
    if debug.EntitiesMessages == nil then debug.EntityMessages = {} end

    mod:AddCallback(ModCallbacks.MC_POST_RENDER, debug.onRender)
end

function debug:onRender()
    if debug.active then
        local f = Font()
        f:Load("font/pftempestasevencondensed.fnt")

        --Screen variables
        for i=1, debug.config.numVar+1, 1 do
            if debug.Variables[i] ~= nil and debug.Variables[i].ressult ~= nil then
                local mesVar = ""
                local pos = Vector(32, 20)

                if debug.Variables[i].name ~= "" then
                    mesVar = debug.Variables[i].name..": "
                end

                local r=1; local g=1; local b=1

                if type(debug.Variables[i].ressult)=="boolean" then
                    if debug.Variables[i].ressult then
                        r=0; g=1; b=0
                    else
                        r=1; g=0; b=0
                    end
                    mesVar = mesVar..Str.bool(debug.Variables[i].ressult)
                else
                    mesVar = mesVar..debug.Variables[i].ressult
                end
                f:DrawStringScaled(mesVar, pos.X, pos.Y+(f:GetLineHeight()/2*(i-1))-(i*2), 0.5, 0.5,KColor(r,g,b,1),0,true)
            end
        end

        --Entity variables
        for index, entity in pairs(debug.EntitiesWithMessages) do

            if not entity:Exists() then
                table.remove(debug.EntitiesWithMessages, index)
            else
                local data = entity:GetData()
                local pos = Game():GetRoom():WorldToScreenPosition(entity.Position)

                for i=1, #data.DebugVariables, 1 do
                    if data.DebugVariables[i] ~= nil and data.DebugVariables[i].ressult ~= nil then
                        local mesVar = ""

                        if data.DebugVariables[i].name ~= "" then
                            mesVar = data.DebugVariables[i].name..": "
                        end

                        local r=1; local g=1; local b=1

                        if type(data.DebugVariables[i].ressult)=="boolean" then
                            if data.DebugVariables[i].ressult then
                                r=0; g=1; b=0
                            else
                                r=1; g=0; b=0
                            end

                            mesVar = mesVar..Str.bool(data.DebugVariables[i].ressult)
                        else
                            mesVar = mesVar..data.DebugVariables[i].ressult
                        end
                        f:DrawStringScaled(mesVar, pos.X - f:GetStringWidth(mesVar)/4, (pos.Y+((f:GetLineHeight()/2*(i-1)))-(i*2)), 0.5, 0.5,KColor(r,g,b,1),0,false)
                    end
                end
            end
        end

        --Mensajes por pantalla
        for i=1, debug.config.numMess+1, 1 do
            local pos = Vector(5, 255-(f:GetLineHeight()/2*(i-1)-(i*2)))
            local mesVar = ""
            local opacity = debug.messageVars.opacity-((1/debug.config.numMess)*(i-1))
            if debug.Messages[i] ~= nil and debug.messageVars.quantity > 0 and i==1 then
                mesVar = ("x"..debug.messageVars.quantity..": "..debug.Messages[i])
            else
                mesVar = debug.Messages[i]
            end
            if debug.Messages[i] ~= nil then f:DrawStringScaled(mesVar, pos.X, pos.Y, 0.5, 0.5, KColor(255,255,255,opacity),0,false) end
        end

        if debug.messageVars.opacity > 0 then
            debug.messageVars.opacity = debug.messageVars.opacity-0.001
        end

        --Mensajes a entidades
        for i=1, #debug.EntityMessages, 1 do
            local opacity = 1

            if debug.EntityMessages[i] ~= nil and debug.EntityMessages[i].displacement < 100 then
                opacity = opacity - debug.EntityMessages[i].displacement/100
            else
                opacity = 0
                if debug.EntityMessages[i] ~= nil then
                    table.remove(debug.EntityMessages, i)
                end
            end

            if debug.EntityMessages[i] ~= nil then
                f:DrawString(debug.EntityMessages[i].message, debug.EntityMessages[i].position.X-f:GetStringWidth(debug.EntityMessages[i].message)/2, debug.EntityMessages[i].position.Y-debug.EntityMessages[i].displacement,KColor(1,1,1,opacity,0,0,0),0,true)
                debug.EntityMessages[i].displacement = (Isaac.GetFrameCount()-debug.EntityMessages[i].initFrame)/3
            end
        end
    end
end

debug.config = {
    numVar = 16,
    numMess = 16,
    entityNumMess = 6
}

debug.messageVars = {
    opacity = 1,
    quantity = 0,
}

debug.logOutput = function(m)
    if m ~= nil then
        Isaac.DebugString(m)
    end
end

debug.output = function (m)
    Isaac.ConsoleOutput(m.."\n")
end

debug.outputVector = function (m,v)
    Isaac.ConsoleOutput(m.."X: "..v.X.." Y: "..v.Y.."\n")
end

debug.setVariable = function(entity, varName, value) --Value, name, entity
    local varArray = debug.Variables
    if entity ~= nil and entity ~= 0 then
        varArray = entity:GetData().DebugVariables
        if varArray == nil then
            entity:GetData().DebugVariables = {}
            table.insert(debug.EntitiesWithMessages, entity)
            varArray = entity:GetData().DebugVariables
        end
    end

    local position = #varArray+1
    for id, var in pairs(varArray) do
        if var.name == varName then
            position = id
        end
    end
    varArray[position]={
        name = varName,
        ressult = value
    }
end

debug.addMessage = function(entity, text)
    --To screen
    if entity == nil or entity == 0 then
        if debug.Messages[1] == text then
            debug.messageVars.quantity = debug.messageVars.quantity+1
        else
            debug.messageVars.quantity = 0
        end

        table.insert(debug.Messages, 1, text)
        debug.messageVars.opacity = 1
        if #debug.Messages > debug.config.numMess then
            table.remove(debug.Messages,debug.config.numMess+1)
        end
    else --To entity
        local newID = #debug.EntityMessages+1
        local newPosition = Game():GetRoom():WorldToScreenPosition(entity.Position)
        local newFrame = Isaac.GetFrameCount()

        debug.EntityMessages[newID] = {
            message = text,
            position = newPosition,
            initFrame = newFrame,
            displacement = 0,
            entity = entity
        }
    end
end

return debug

File: src/debug.ts
/* eslint-disable max-classes-per-file */
import { ModCallback } from "isaac-typescript-definitions";
import {
  VectorZero,
  arrayRemoveIndex,
  game,
  getEntityFromPtrHash,
} from "isaacscript-common";
import { mod } from "./mod";

class ScreenMessageClass {
  messages: string[] = [];
  opacity = 1;
  lastMessageQuantity = 0;
}

class EntityMessageClass {
  displacement = 0;
  message = "";
  position: Vector = VectorZero;
  initFrame = 0;
}

class DebugClass {
  active = true;
  screen: ScreenMessageClass = new ScreenMessageClass();
  position: EntityMessageClass[] = [];
  entities: Map<PtrHash, Map<string, unknown>> = new Map<
    PtrHash,
    Map<string, unknown>
  >();

  variables: Map<string, unknown> = new Map<string, unknown>();
}

const debug = new DebugClass();

const preconf = {
  numScreenMess: 3,
  numEntityMess: 3,
};

export function init(): void {
  mod.AddCallback(ModCallback.POST_RENDER, renderDebug);
}

/**
 * Set a debug variable on a entity or on the top left of the screen.
 *
 * @param varName This need to be unique to every entity or the screen.
 * @param value Variable that will be displayed.
 * @param ent Optional. Entity where variable will be displayed.
 */
export function setVariable(
  varName: string,
  value: unknown,
  ent?: Entity,
): void {
  if (ent !== undefined) {
    // To entity
    const ID = GetPtrHash(ent);

    if (ent.Exists()) {
      // Ensure that is alive.
      let variables: Map<string, unknown> = new Map<string, unknown>();
      variables.set(varName, value);
      if (debug.entities.get(ID) === undefined) {
        // New entity.
        debug.entities.set(ID, variables);
      } else {
        // Saved entity.
        const thisData = debug.entities.get(ID);
        if (thisData !== undefined) {
          thisData.set(varName, value);
        }
      }
    } else if (debug.entities.get(ID) !== undefined) {
      debug.entities.delete(ID);
    }
  } else {
    // To screen.
    debug.variables.set(varName, value);
  }
}

/**
 * Adds a message that will be rendered at the bottom right of the screen or on a entity position.
 *
 * @param text Message that will be displayed.
 * @param entPos Optional. Position where message will be displayed, use entity to take its
 *               position.
 */
export function addMessage(text: string, entPos?: Entity): void {
  // To screen
  if (entPos === undefined) {
    // Add lastMessageQuantity if it is a repeated message.
    if (debug.screen.messages[1] === text) {
      debug.screen.lastMessageQuantity++;
    } else {
      debug.screen.lastMessageQuantity = 0;
    }

    debug.screen.messages.push(text);
    debug.screen.opacity = 1;

    // Limit message num
    if (debug.screen.messages.length > preconf.numScreenMess) {
      arrayRemoveIndex(debug.screen.messages, preconf.numScreenMess + 1);
    }
  } else {
    // To entity
    const newEntityMess = new EntityMessageClass();

    newEntityMess.message = text;
    newEntityMess.initFrame = Isaac.GetFrameCount();
    newEntityMess.displacement = 0;

    newEntityMess.position = game
      .GetRoom()
      .WorldToScreenPosition(entPos.Position);

    debug.position.push(newEntityMess);
  }
}

/** Used on PostRenderCallback, render debug messages and variables. */
function renderDebug() {
  if (debug.active) {
    const f = Font();
    f.Load("font/pftempestasevencondensed.fnt");

    // Screen messages
    let message = "";
    if (debug.screen.messages.length > 0) {
      for (let i = 1; i < preconf.numScreenMess + 1; i++) {
        if (debug.screen.messages[i] !== undefined) {
          if (i === 1 && debug.screen.lastMessageQuantity > 0) {
            message = `x${debug.screen.lastMessageQuantity}: ${debug.screen.messages[i]}`;
          } else {
            message = `${debug.screen.messages[i]}`;
          }
          Isaac.RenderText(
            message,
            5,
            255 - 11 * (i - 1),
            1,
            1,
            1,
            debug.screen.opacity - (1 / preconf.numScreenMess) * (i - 1),
          );
        }
      }

      if (debug.screen.opacity > 0) {
        debug.screen.opacity -= 0.003;
      }
    }

    // Position messages
    const numEntities = debug.position.length;
    if (numEntities > 0) {
      for (let i = 0; i < numEntities + 1; i++) {
        let opacity = 1;
        const entityMess = debug.position[i];

        if (entityMess !== undefined) {
          if (entityMess.displacement < 100) {
            opacity -= entityMess.displacement / 100;
            f.DrawString(
              entityMess.message,
              entityMess.position.X - f.GetStringWidth(entityMess.message) / 2,
              entityMess.position.Y - entityMess.displacement,
              KColor(1, 1, 1, opacity),
              0,
              true,
            );
            entityMess.displacement =
              (Isaac.GetFrameCount() - entityMess.initFrame) / 3;
          } else {
            opacity = 0;
            debug.position = arrayRemoveIndex(debug.position, i);
          }
        }
      }
    }

    let r = 0;
    let g = 0;
    let b = 0;

    // Screen variables
    if (debug.variables.size > 0) {
      let messVar = "";
      let i = 1;
      for (const [name, value] of debug.variables) {
        if (name !== "") {
          messVar = `${name}: `;
        }

        if (typeof value === "boolean") {
          if (value) {
            r = 0;
            g = 1;
            b = 0;
          } else {
            r = 1;
            g = 0;
            b = 0;
          }
        }
        messVar += `${value}`;
        Isaac.RenderText(messVar, 32, 28 + 11 * (i - 1), r, g, b, 1);
        i++;
      }
    }

    // Entity variables
    if (debug.entities.size > 0) {
      for (const [ptr, variables] of debug.entities) {
        // Remove entity if it is not alive.
        const thisEnt = getEntityFromPtrHash(ptr);
        if (thisEnt === undefined || !thisEnt.Exists()) {
          debug.entities.delete(ptr);
        } else {
          const pos = game.GetRoom().WorldToScreenPosition(thisEnt.Position);

          let i = 1;
          for (const [name, value] of variables) {
            let messVar = "";

            if (name !== "") {
              messVar = `${name}: `;
            }

            if (typeof value === "boolean") {
              if (value) {
                r = 0;
                g = 1;
                b = 0;
              } else {
                r = 1;
                g = 0;
                b = 0;
              }
            }

            messVar += `${value}`;
            f.DrawString(
              messVar,
              pos.X - f.GetStringWidth(messVar) / 2,
              pos.Y + f.GetLineHeight() * (i - 1),
              KColor(r, g, b, 1),
              0,
              true,
            );
            i++;
          }
        }
      }
    }
  }
}
